File: SteamDroid/src/com/flyingpie/steamdroid/adapter/ChatAdapter.java
package com.flyingpie.steamdroid.adapter;

import java.util.ArrayList;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.TextView;

import com.flyingpie.steamdroid.R;
import com.flyingpie.steamdroid.api.ChatMessage;
import com.flyingpie.steamdroid.api.State;

public class ChatAdapter extends BaseAdapter {

	private ArrayList<ChatMessage> messages;
	private LayoutInflater inflater;
	private Context context;
	
	public ChatAdapter(Context context, ArrayList<ChatMessage> messages)
	{
		this.messages = messages;
		this.inflater = LayoutInflater.from(context);
		this.context = context;
	}
	
	public void add(ChatMessage message)
	{
		messages.add(message);
	}
	
	public void remove(ChatMessage message)
	{
		messages.remove(message);
	}
	
	public void clear()
	{
		messages.clear();
	}
	
	@Override
	public int getCount() {
		return messages.size();
	}

	@Override
	public Object getItem(int position) {
		return messages.get(position);
	}

	@Override
	public long getItemId(int position) {
		return position;
	}

	@Override
	public View getView(int position, View convertView, ViewGroup parent) {
		ChatViewHolder holder;
		
		if(convertView == null)
		{
			convertView = inflater.inflate(R.layout.adapter_chat, null);
			holder = new ChatViewHolder();
			holder.textName = (TextView)convertView.findViewById(R.id.chatName);
			holder.textMessage = (TextView)convertView.findViewById(R.id.chatMessage);
			
			convertView.setTag(holder);
		}
		else
		{
			holder = (ChatViewHolder)convertView.getTag();
		}
		
		ChatMessage message = messages.get(position);
		
		holder.textName.setText(message.getSender().getName());
		holder.textMessage.setText(message.getMessage());
		
		int color = R.color.stateOnline;
		
		if(message.getSender().getState().equals(State.getStateOffline()))
		{
			color = R.color.stateOffline;
		}
		
		if(message.getSender().isPlaying())
		{
			color = R.color.stateInGame;
		}
		
		color = context.getResources().getColor(color);
		
		holder.textName.setTextColor(color);
		
		return convertView;
	}

	static class ChatViewHolder
	{
		TextView textName;
		TextView textMessage;
	}
}

File: SteamDroid2/SteamDroid2/Adapters/ChatAdapter.cs
using System.Collections.Generic;
using System.Linq;
using Android.Content;
using Android.Views;
using Android.Widget;

using SteamKit2;
using SteamDroid.Api;

namespace SteamDroid
{
    public class ChatAdapter : BaseAdapter, ICallbackHandler
    {
        private Context context;
        private List<ChatMessage> messages;
        private LayoutInflater inflater;
        
        private Friend friend;
        
        public ChatAdapter(Friend friend, Context context)
        {
            this.context = context;
            messages = new List<ChatMessage>();
            inflater = LayoutInflater.FromContext(context);
            this.friend = friend;
            
            SteamService.GetClient().AddHandler(this);
        }
        
        public override int Count
        {
            get { return messages.Count; }
        }
        
        public void Add(ChatMessage message)
        {
            messages.Add(message);
        }
        
        public ChatMessage GetMessageAt(int position)
        {
            return messages.ElementAt(position);
        }
        
        public override Java.Lang.Object GetItem(int position)
        {
            return position;
        }
        
        public override long GetItemId(int position)
        {
            return position;
        }
        
        public override View GetView(int position, View view, ViewGroup parent)
        {
            ViewHolder holder;
            
            if(view == null)
            {
                view = inflater.Inflate(Resource.Layout.AdapterChat, null);
                holder = new ViewHolder();
                holder.TextName = view.FindViewById<TextView>(Resource.Id.ChatName);
                holder.TextMessage = view.FindViewById<TextView>(Resource.Id.ChatMessage);
                
                view.Tag = holder;
            }
            
            holder = (ViewHolder)view.Tag;
            
            ChatMessage message = GetMessageAt(position);
            
            holder.TextName.Text = message.Friend.Name;
            holder.TextMessage.Text = message.Message;
            
            int color = Resource.Color.StateOnline;
            
            if(message.Friend.State == EPersonaState.Offline)
            {
                color = Resource.Color.StateOffline;
            }
            else if(message.Friend.GameName.Length > 0)
            {
                color = Resource.Color.StateInGame;
            }
            
            color = context.Resources.GetColor (color);
            
            holder.TextName.SetTextColor(color);
            
            return view;
        }
        
        public void HandleCallback(CallbackMsg msg)
        {
            if(msg.IsType<SteamFriends.FriendMsgCallback>())
            {
                SteamFriends.FriendMsgCallback callback = (SteamFriends.FriendMsgCallback)msg;
                
                if(callback.EntryType == EChatEntryType.ChatMsg && callback.Sender == friend.SteamId)
                {
                    ChatMessage message = new ChatMessage(friend, callback.Message);
                    Add(message);
                    NotifyDataSetChanged();
                }
            }
        }
        
        class ViewHolder : Java.Lang.Object
        {
            public TextView TextName { get; set; }
            public TextView TextMessage { get; set; }
        }
    }
}

